src/fir.cu

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) 2017, The Bifrost Authors. All rights reserved.
+ * Copyright (c) 2017, The University of New Mexico. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ * * Neither the name of The Bifrost Authors nor the names of its
+ *   contributors may be used to endorse or promote products derived
+ *   from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <bifrost/fir.h>
+#include "assert.hpp"
+#include "utils.hpp"
+#include "workspace.hpp"
+#include "cuda.hpp"
+#include "trace.hpp"
+#include "Complex.hpp"
+
+//#include <limits>
+
+#include <math_constants.h> // For CUDART_NAN_F
+#include <thrust/device_vector.h>
+
+#include <vector>
+#include <map>
+#include <string>
+
+// HACK TESTING
+#include <iostream>
+using std::cout;
+using std::endl;
+
+template<typename InType, typename OutType>
+__global__ void fir_kernel(int                   ncoeff,
+                           int                   decim, 
+                           int                   ntime, 
+                           int                   nstand,
+                           double*               coeffs,
+                           Complex64*            state0,
+                           Complex64*            state1,
+                           InType*               d_in,
+                           OutType* __restrict__ d_out) {
+	int t0 = threadIdx.x + (blockIdx.x + blockIdx.y*gridDim.x)*blockDim.x;
```

Comment:
I initially went with this approach on ADP because the first usage case had ntime >> nstand * npol.  Is there a good thread/block division rubric that can be used to automatically tune the setup?  Also, is there another way to deal with one dimension having a large number of elements?